src/runtime/var.rs

```diff
@@ -1,46 +1,83 @@
 use crate::{Commit, Key};
-use parking_lot::Mutex;
-use std::{sync::Arc, task::Waker};
+use parking_lot::{Mutex, RwLock};
+use std::sync::Arc;
+
+use super::{Revision, RevisionControlSystem};
 
 /// The underlying container of state variables. Vends copies of the latest
 /// [`Commit`] for [`Key`]s.
 pub(crate) struct Var<State> {
     current: Commit<State>,
     id: topo::CallId,
-    pending: Option<Commit<State>>,
-    waker: Waker,
+    // can only contain commit from current revision
+    // make proper checks inside methods!
+    pending: Option<(Revision, Commit<State>)>,
+    rcs: Arc<RwLock<RevisionControlSystem>>,
 }
 
 impl<State> Var<State> {
-    pub fn new(id: topo::CallId, waker: Waker, inner: State) -> Arc<Mutex<Self>> {
+    pub fn new(
+        id: topo::CallId,
+        rcs: Arc<RwLock<RevisionControlSystem>>,
+        inner: State,
+    ) -> Arc<Mutex<Self>> {
         let current = Commit { id, inner: Arc::new(inner) };
-        Arc::new(Mutex::new(Var { id, current, waker, pending: None }))
+        Arc::new(Mutex::new(Var { id, current, rcs, pending: None }))
     }
 
     /// Attach this `Var` to its callsite, performing any pending commit and
     /// returning the resulting latest commit.
     pub fn root(var: Arc<Mutex<Self>>) -> (Commit<State>, Key<State>) {
         let (id, commit_at_root) = {
             let mut var = var.lock();
-            if let Some(pending) = var.pending.take() {
-                var.current = pending;
+            // This function is always called within it's context
+            let current = Revision::current();
+
+            // Replace current commit with pending commit if it is from the past revision
+            match var.pending.take() {
+                Some((revision, commit)) if revision < current => var.current = commit,
+                still_pending => var.pending = still_pending,
             }
```

Comment:
Can you extract this to a method like `Var::maybe_commit_pending`?

Reply:
Cannot bc it would mutably borrow a whole `Var`. I couldn't get it to work.